ext/openssl/openssl.c

```diff
@@ -4228,180 +4228,280 @@ static EVP_PKEY *php_openssl_pkey_init_dh(zval *data, bool *is_private)
+	} else {
+		OPENSSL_PKEY_SET_BN(data, p);
+		OPENSSL_PKEY_SET_BN(data, a);
+		OPENSSL_PKEY_SET_BN(data, b);
+		OPENSSL_PKEY_SET_BN(data, order);
+		OPENSSL_PKEY_SET_BN(data, g_x);
+		OPENSSL_PKEY_SET_BN(data, g_y);
+		OPENSSL_PKEY_SET_BN(data, cofactor);
+
+		if (!(p && a && b && order)) {
+			php_error_docref(NULL, E_WARNING, "Missing params: need \"p\", \"a\", \"b\", \"order\" ");
+			goto cleanup;
+		}
 
-	if (d) {
-		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_PRIV_KEY, d);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_P, p);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_A, a);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_B, b);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_ORDER, order);
+		OSSL_PARAM_BLD_push_utf8_string(bld, OSSL_PKEY_PARAM_EC_FIELD_TYPE, SN_X9_62_prime_field, 0);
 
-		pnt = EC_POINT_new(group);
-		if (!pnt || !EC_POINT_mul(group, pnt, d, NULL, NULL, NULL)) {
+		if(!(group = EC_GROUP_new_curve_GFp(p, a, b, bctx))) {
 			goto cleanup;
 		}
-	} else if (x && y) {
-		/* OpenSSL does not allow setting EC_PUB_X/EC_PUB_Y, so convert to encoded format. */
-		pnt = EC_POINT_new(group);
-		if (!pnt || !EC_POINT_set_affine_coordinates(group, pnt, x, y, NULL)) {
+		
+		point_g = EC_POINT_new(group);
+
+		if (generator_zv && Z_TYPE_P(generator_zv) == IS_STRING && Z_STRLEN_P(generator_zv) > 0) {
+			EC_POINT_oct2point(group, point_g, Z_STRVAL_P(generator_zv), Z_STRLEN_P(generator_zv), bctx);
+			OSSL_PARAM_BLD_push_octet_string(
+				bld, OSSL_PKEY_PARAM_EC_GENERATOR, Z_STRVAL_P(generator_zv), Z_STRLEN_P(generator_zv));
+		} else if(g_x && g_y) {
+			if (!EC_POINT_set_affine_coordinates(group, point_g, g_x, g_y, bctx)) {
+				goto cleanup;
+			}
+
+			size_t point_g_buf_len =
+				EC_POINT_point2buf(group, point_g, POINT_CONVERSION_COMPRESSED, &point_g_buf, bctx);
+			if (!point_g_buf_len) {
+				goto cleanup;
+			}
+
+			OSSL_PARAM_BLD_push_octet_string(bld, OSSL_PKEY_PARAM_EC_GENERATOR, point_g_buf, point_g_buf_len);
+		} else {
+			php_error_docref(
+				NULL, E_WARNING, "Missing params: need \"generator\" or \"g_x\" and \"g_y\" ");
 			goto cleanup;
 		}
+
+		if (seed_zv && Z_TYPE_P(seed_zv) == IS_STRING && Z_STRLEN_P(seed_zv) > 0) {
+			OSSL_PARAM_BLD_push_octet_string(
+				bld, OSSL_PKEY_PARAM_EC_SEED, Z_STRVAL_P(seed_zv), Z_STRLEN_P(seed_zv));
+		}
+
+		if (cofactor) {
+			OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_COFACTOR, cofactor);
+			EC_GROUP_set_generator(group, point_g, order, cofactor);
+		} else {
+			EC_GROUP_set_generator(group, point_g, order, BN_value_one());
+		}
 	}
 
-	if (pnt) {
-		size_t pnt_oct_len =
-			EC_POINT_point2buf(group, pnt, POINT_CONVERSION_COMPRESSED, &pnt_oct, NULL);
-		if (!pnt_oct_len) {
-			goto cleanup;
+	/* custom params not supported with SM2, SKIP */
+	if (!curve_name_zv ||
+		(Z_TYPE_P(curve_name_zv) == IS_STRING &&
+		OPENSSL_strcasecmp(Z_STRVAL_P(curve_name_zv), "SM2") != 0)){
```

Comment:
This should probably go under the condition https://github.com/php/php-src/pull/9991/files#diff-b08a60bcfd312a9899392bfd4c5ba3427f9be6aa43c37808064bc60f7c751c9bR4376 .

Reply:
Just skip the logic which import the keypairs to OSSL_PARAM under the condition "curve_name" is exactly set to "SM2", because the operations to generate keypairs with SM2's priv/pub params ("x", "y", "d")  cannnot be executed (although other curves behave differently). 

In order to be compatible with OpenSSL versions below 3.0 as much as possible, considering the particularity of SM2, we only need to judge the curve defined by the curve name for the time being.

Reply:
Sorry, it seems that I have misunderstood the meaning of this document. [SM2 Key Changes](https://github.com/openssl/openssl/blob/openssl-3.0.7/doc/man7/migration_guide.pod#sm2-key-changes)

It seems to express that SM2 context must be created directly before generating SM2 key pair, not EC.

SM2 can only import keys by creating an SM2 context first. Therefore, if SM2 curves are created through explicit parameters, I can directly switch to the SM2 context to avoid the problem that SM2 keys cannot be imported.

However, it can only support OpenSSL versions above 3.0. Because the standard for generating SM2 key pairs below 3.0 is inconsistent with it, it should be OK to do so.

I have submitted changes, please review, thank you so much.